File: backend/app/schemas/rubric.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator, model_validator


class CriterionInput(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = ""
    weight: float = Field(gt=0, le=100)
    order_index: int = 0


class CriterionOut(CriterionInput):
    id: int

    model_config = {"from_attributes": True}


class RubricCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    disagreement_threshold: float = Field(default=1.5, gt=0, le=10)
    criteria: list[CriterionInput] = Field(min_length=1)

    @field_validator("criteria")
    @classmethod
    def weights_sum_to_100(cls, criteria: list[CriterionInput]) -> list[CriterionInput]:
        total = round(sum(c.weight for c in criteria), 2)
        if total != 100:
            raise ValueError(f"Criterion weights must total 100%, got {total}%")
        return criteria


class RubricUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    disagreement_threshold: float | None = Field(default=None, gt=0, le=10)
    is_active: bool | None = None
    criteria: list[CriterionInput] | None = None

    @model_validator(mode="after")
    def weights_sum_to_100(self) -> "RubricUpdate":
        if self.criteria is not None:
            total = round(sum(c.weight for c in self.criteria), 2)
            if total != 100:
                raise ValueError(f"Criterion weights must total 100%, got {total}%")
        return self


class RubricOut(BaseModel):
    id: int
    name: str
    is_active: bool
    disagreement_threshold: float
    created_at: datetime
    criteria: list[CriterionOut]

    model_config = {"from_attributes": True}

File: workers/src/schemas/rubric.ts
import { z } from "zod";

/** Mirrors backend/app/schemas/rubric.py. */
export const criterionInputSchema = z.object({
  name: z.string().min(1).max(120),
  description: z.string().default(""),
  weight: z.number().gt(0).lte(100),
  order_index: z.number().int().default(0),
});

/** Pydantic rounded the sum to 2dp before comparing, so float drift does not reject. */
function weightsTotal(criteria: { weight: number }[]): number {
  return Math.round(criteria.reduce((sum, c) => sum + c.weight, 0) * 100) / 100;
}

function weightsTotal100(criteria: { weight: number }[]): boolean {
  return weightsTotal(criteria) === 100;
}

/**
 * The message is rendered exactly as Pydantic did, because the frontend prints `detail`
 * verbatim. `weight` is a Python float so an integral total renders as "90.0", except for
 * an empty list where `sum([])` returns the int 0 and renders as "0". Pydantic v2 also
 * prefixes any ValueError raised inside a validator with "Value error, ".
 */
function weightsMessage(criteria: { weight: number }[]): string {
  const total = weightsTotal(criteria);
  const rendered =
    criteria.length === 0 ? "0" : Number.isInteger(total) ? total.toFixed(1) : String(total);
  return `Value error, Criterion weights must total 100%, got ${rendered}%`;
}

export const rubricCreateSchema = z
  .object({
    name: z.string().min(1).max(120),
    disagreement_threshold: z.number().gt(0).lte(10).default(1.5),
    criteria: z.array(criterionInputSchema).min(1),
  })
  .superRefine((data, ctx) => {
    if (!weightsTotal100(data.criteria)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["criteria"], message: weightsMessage(data.criteria) });
    }
  });

/**
 * `criteria` accepts an explicit null: RubricUpdate declared it `list | None`, so a null
 * validated and `if data.criteria is not None` skipped the replacement, leaving the
 * existing criteria intact and returning 200.
 */
export const rubricUpdateSchema = z
  .object({
    name: z.string().min(1).max(120).optional(),
    disagreement_threshold: z.number().gt(0).lte(10).optional(),
    is_active: z.boolean().optional(),
    criteria: z.array(criterionInputSchema).nullable().optional(),
  })
  .superRefine((data, ctx) => {
    if (data.criteria != null && !weightsTotal100(data.criteria)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["criteria"], message: weightsMessage(data.criteria) });
    }
  });

export type CriterionInput = z.infer<typeof criterionInputSchema>;
